src/Perlang.Interpreter/PerlangInterpreter.cs

```diff
@@ -1123,41 +1364,338 @@ public VoidObject VisitWhileStmt(Stmt.While stmt)
                     }
 
                 case PERCENT:
                     CheckNumberOperands(expr.Operator, left, right);
-                    return leftNumber % rightNumber;
+
+                    // Bang operator (!) is safe because of the CheckNumberOperands() call above.
+                    if (left is float or double && right is float or double)
+                    {
+                        double leftNumber = leftConvertible!.ToDouble(CultureInfo.InvariantCulture);
+                        double rightNumber = rightConvertible!.ToDouble(CultureInfo.InvariantCulture);
+
+                        return leftNumber % rightNumber;
+                    }
+                    else if (left is int or long && right is float or double)
+                    {
+                        long leftNumber = leftConvertible!.ToInt64(CultureInfo.InvariantCulture);
+                        double rightNumber = rightConvertible!.ToDouble(CultureInfo.InvariantCulture);
+
+                        return leftNumber % rightNumber;
+                    }
+                    else if (left is int && right is int)
+                    {
+                        int leftNumber = leftConvertible!.ToInt32(CultureInfo.InvariantCulture);
+                        int rightNumber = rightConvertible!.ToInt32(CultureInfo.InvariantCulture);
+
+                        return leftNumber % rightNumber;
+                    }
+                    else if (left is long && right is long)
+                    {
+                        long leftNumber = leftConvertible!.ToInt64(CultureInfo.InvariantCulture);
+                        long rightNumber = rightConvertible!.ToInt64(CultureInfo.InvariantCulture);
+
+                        return leftNumber % rightNumber;
+                    }
+                    else if (left is BigInteger && right is BigInteger)
+                    {
+                        var leftBigInt = (BigInteger)left;
+                        var rightBigInt = (BigInteger)right;
+
+                        return leftBigInt % rightBigInt;
+                    }
+                    else if (left is int or long && right is BigInteger)
+                    {
+                        long leftNumber = leftConvertible!.ToInt64(CultureInfo.InvariantCulture);
+                        var rightBigInt = (BigInteger)right;
+
+                        return leftNumber % rightBigInt;
+                    }
+                    else if (left is BigInteger && right is int or long)
+                    {
+                        var leftBigInt = (BigInteger)left;
+                        long rightNumber = rightConvertible!.ToInt64(CultureInfo.InvariantCulture);
+
+                        return leftBigInt % rightNumber;
+                    }
+                    else
+                    {
+                        throw new RuntimeError(expr.Operator, $"Operands must be numbers, not {StringifyType(left)} and {StringifyType(right)}");
+                    }
```

Comment:
:point_up: This (all the changes in this file) was quite a bit of work and I'm not overly pleased with the end result. The problem is that for code that deals with dynamic data of multiple types (`object`), it's really, really hard to do things like this in a clean way.

Possible alternative approach: [use generic math](https://devblogs.microsoft.com/dotnet/preview-features-in-net-6-generic-math/), which is available as a preview in .NET 6. This would possibly _work_ but was nontrivial when I gave it a try. If we would make `Expr.Literal` be a generic type `Expr.Literal<T : INumber>`, it would probably work for constant _numeric_ expressions (but how do we handle literal strings in that case?). We would also need to make it work for `Expr.Identifier`-based expressions as well (i.e. variable references/function calls), and the way our current scanner & parser works, this is simply not doable - the type of an expression is compiled much later than the expression tree is being constructed.

So the TL;DR for generic math is that it's not really helpful for us, AFAIK. (Feel free to prove me wrong, I would really love it if it was doable.)